File: Solution-Python/Substring_30.py
from collections import defaultdict

class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        res = []
        wl = len(words[0])
        freq = defaultdict(int)
        for w in words:
            freq[w] += 1
        for i in range(wl):
            start = i
            count = len(words)
            seen = defaultdict(int)
            for end in range(start + wl, len(s) + 1, wl):
                tail = s[end - wl: end]
                if tail in freq:
                    seen[tail] += 1
                    if seen[tail] <= freq[tail]:
                        count -= 1
                    if end - start == wl * len(words):
                        if count == 0:
                            res.append(start)
                        head = s[start: start + wl]
                        seen[head] -= 1
                        if seen[head] < freq[head]:
                            count += 1
                        start += wl
                else:
                    start = end
                    count = len(words)
                    seen = defaultdict(int)
        return res

File: Solution-Go/Substring_30.go
func findSubstring(s string, words []string) []int {
    wl := len(words[0])
    res := []int{}
    freq := make(map[string]int)
    for _, word := range words {
        freq[word] += 1
    }
    for i := 0; i < wl; i++ {
        start := i
        count := len(words)
        seen := make(map[string]int)
        for end := start + wl; end <= len(s); end += wl {
            tail := s[end - wl: end]
            if _, ok := freq[tail]; ok {
                if seen[tail]++; seen[tail] <= freq[tail] {
                    count--
                }
                if end - start == wl * len(words) {
                    if count == 0 {
                        res = append(res, start)
                    }
                    head := s[start: start + wl]
                    if seen[head]--; seen[head] < freq[head] {
                        count++
                    }
                    start += wl
                }
            } else {
                start = end
                count = len(words)
                seen = make(map[string]int)
            }
        }
    }
    return res
}